### PR TITLE
Optimizer: improve the load-copy-to-borrow optimization and implement it in swift

### PR DESCRIPTION
**Pull request by eeckstein**
The optimization replaces a `load [copy]` with a `load_borrow` if possible.

```
  %1 = load [copy] %0
  // no writes to %0
  destroy_value %1
```
->
```
  %1 = load_borrow %0
  // no writes to %0
  end_borrow %1
```

The new implementation uses alias-analysis (instead of a simple def-use walk), which is much more powerful.
Also, implement load-borrow-immutability checkin in the swift verifier.

rdar://115315849

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by eeckstein** [comment]
@swift-ci test

File: SwiftCompilerSources/Sources/Optimizer/PassManager/Context.swift [hunk 1]
```diff
@@ -422,6 +422,14 @@ struct FunctionPassContext : MutatingContext {
 
       return buildFn(specializedFunction, nestedFunctionPassContext)
   }
+
+  /// Makes sure that the lifetime of `value` ends at all control flow paths, even in dead-end blocks.
+  /// Inserts destroys in dead-end blocks if those are missing.
+  func completeLifetime(of value: Value) {
```
Comment (nate-chandler) [inline SwiftCompilerSources/Sources/Optimizer/PassManager/Context.swift:428]: Is now a good time to add a parameter to control whether lifetimes end at the availability or lifetime boundary?

File: SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LoadCopyToBorrowOptimization.swift [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+  //     ...          // no destroy of %1!
+  //   unreachable    // the lifetime of %1 ends here
+  //
+  context.completeLifetime(of: loadedValue)
+
+  if !ownershipUses.collectUses(of: loadedValue) {
+    return
+  }
+
+  var liferange = InstructionWorklist(context)
```
Comment (nate-chandler) [inline SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LoadCopyToBorrowOptimization.swift:90]: Nit: `liferange` -> `liverange` (or `lifetime`) throughout.

File: SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LoadCopyToBorrowOptimization.swift [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+  // missing at the `unreachable`. In such a case the dataflow would _not_ visit potential writes to the load's
+  // memory location. In the following example, the `load [copy]` must not be converted to a `load_borrow`:
+  //
+  //   %1 = load [copy] %0
+  //     ...
+  //   store %2 to %0
+  //     ...          // no destroy of %1!
+  //   unreachable    // the lifetime of %1 ends here
+  //
+  context.completeLifetime(of: loadedValue)
```
Comment (nate-chandler) [inline SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LoadCopyToBorrowOptimization.swift:84]: Generally, lifetimes need to be completed inside out/bottom up (as in `SILGenCleanup::completeOSSALifetimes`).  I haven't come up with an example where that would be a problem in this usage, but I'm not confident it's not.

File: SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LoadCopyToBorrowOptimization.swift [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+          worklist.pushIfNotVisited(contentsOf: forwardingInst.forwardedResults)
+        default:
+          return false
+        }
+      }
+    }
+    return true
+  }
+
+  private mutating func findNonDestroyingLiferangeExists(of forwardingInst: ForwardingInstruction) {
```
Comment (nate-chandler) [inline SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LoadCopyToBorrowOptimization.swift:206]: `findNonDestroyingLiferangeExists` -> `findNonDestroyingLiferangeExits`

File: SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LoadCopyToBorrowOptimization.swift [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+  if load.loadOwnership != .copy {
+    return
+  }
+
+  var ownershipUses = OwnershipUses(context)
+  defer { ownershipUses.deinitialize() }
+
+  // Handle the special case if the `load` is immediately followed by a single `move_value`.
+  // Note that `move_value` is _not_ a forwarding instruction.
+  // In this case we have to preserve the move's flags by inserting a `begin_borrow` with the same flags.
```
Comment (nate-chandler) [inline SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LoadCopyToBorrowOptimization.swift:58]: Would this portion of the comment be most useful in `replaceWithLoadBorrow(liferange:ownershipUses:)` near
```
     if let moveInst = singleMoveValueUser {
```
?

**Comment by eeckstein** [comment]
@swift-ci test macos

**Review by nate-chandler — APPROVED** [review]
LGTM, modulo the question of completing lifetimes inside out.

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test

**Review by atrick — COMMENTED** [review]
Observations and future suggestions...

--- Liverange::collectUses ---

`InteriorUseWalker::init` does not actually need `visitInnerUses: Bool` because it could just have a `visitInnerUses` entry point that does:
```
innerScopeHandler = { return visitUsesOfOuter(value: $0) }
```

`Liverange::collectUses` is just doing what I would call "forwarding-liveness". We could standardize that: basically a combination of `computeLinearLiveness` and `ForwardingDefUseWalker`.

`findAdditionalUsesInDeadEndBlocks` can be eliminated with complete OSSA... ok, I just noticed a TODO says the same thing. But until then, it seems like it might be as efficient to call `completeLife(of:)` first for each owned values, then simply look for `endsLifetime` uses. Note that you do need to handle `ExtendLifetimeInst`, just like `computeLinearLiveness` does.

--- LoadBorrowInst::verify ---

Small note on load-borrow-immutability verification:

At some point, I would like to centralize the definition of legal address uses in one place. So `isMutatedAddress` would be alongside `AddressUseVisitor::classifyAddress`. Also note that, unlike the incorrectly named `AddressDefUseWalker::leafUse`, `AddressUseVisitor::leafAddressUse` is actually a leaf use, so you could probably just check `mayWrite` instead of switching over opcodes.

Then verification should be improved so it does not ignore "unknown" uses (although in this particular verification, we always need to ignore *known escaping* uses).

**Comment by eeckstein** [comment]
> InteriorUseWalker::init does not actually need visitInnerUses: Bool because it could just have a visitInnerUses entry point that does:

👍. Though the drawback is that this is an escaping closure and needs memory allocation for its context. It's potentially called many times (for each load). So this might affect compile time.

> so you could probably just check mayWrite instead of switching over opcodes.

That doesn't work because `mayWrite` is less precise than what the load-copy-to-borrow optimization does using alias analysis. It would result in false alarms.

**Pull request merged**